cpp/src/parquet/arrow/writer.cc

```diff
@@ -406,57 +402,7 @@ class FileWriterImpl : public FileWriter {
     }
 
     // Max number of rows allowed in a row group.
-    const int64_t max_row_group_length = this->properties().max_row_group_length();
-
-    if (row_group_writer_ == nullptr || !row_group_writer_->buffered() ||
-        row_group_writer_->num_rows() >= max_row_group_length) {
-      RETURN_NOT_OK(NewBufferedRowGroup());
-    }
-
-    auto WriteBatch = [&](int64_t offset, int64_t size) {
-      std::vector<std::unique_ptr<ArrowColumnWriterV2>> writers;
-      int column_index_start = 0;
-
-      for (int i = 0; i < batch.num_columns(); i++) {
-        ChunkedArray chunked_array{batch.column(i)};
-        ARROW_ASSIGN_OR_RAISE(
-            std::unique_ptr<ArrowColumnWriterV2> writer,
-            ArrowColumnWriterV2::Make(chunked_array, offset, size, schema_manifest_,
-                                      row_group_writer_, column_index_start));
-        column_index_start += writer->leaf_count();
-        if (arrow_properties_->use_threads()) {
-          writers.emplace_back(std::move(writer));
-        } else {
-          RETURN_NOT_OK(writer->Write(&column_write_context_));
-        }
-      }
-
-      if (arrow_properties_->use_threads()) {
```

Comment:
Moving the async logic into `WriteTable` may cause a potential concurrent issue mentioned by @westonpace in this [comment](https://github.com/apache/arrow/pull/33656#pullrequestreview-1250353638).

Reply:
oops... It's really amazing

Reply:
You may also add similar comment to WriteTable

```cpp
  /// WARNING: If you are writing multiple files in parallel in the same
  /// executor, deadlock may occur if ArrowWriterProperties::use_threads
  /// is set to true to write columns in parallel. Please disable use_threads
  /// option in this case.
```

Reply:
This is added in declaration